Migrate reciprocal from the TH to Aten (CUDA)

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30422 Remove exp10 in TH (unused)
* #30329 Remove many unused bfloat16 functions in TH
* **#30176 Migrate reciprocal from the TH to Aten (CUDA)**

Benchmark (Debian buster, Release build, P400, turbo off):

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.reciprocal(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.reciprocal(a); torch.cuda.synchronize()', setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")', number=t))
```

Before:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.37665776500034553
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.3755079170005047
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45436807200167095
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7375995410002361
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.989992897000775
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7109333839998726
```

After:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.3880071029998362
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.38418854500014277
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45793338999919797
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7040607610006191
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.9805948499997612
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7015677830004279
```

Differential Revision: [D18764186](https://our.internmc.facebook.com/intern/diff/D18764186)

xuhdev committed:
Migrate reciprocal from the TH to Aten (CUDA)

Benchmark (Debian buster, Release build, P400, turbo off):

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.reciprocal(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.reciprocal(a); torch.cuda.synchronize()', setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")', number=t))
```

Before:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.37665776500034553
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.3755079170005047
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45436807200167095
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7375995410002361
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.989992897000775
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7109333839998726
```

After:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.3880071029998362
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.38418854500014277
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45793338999919797
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7040607610006191
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.9805948499997612
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7015677830004279
```

[ghstack-poisoned]

xuhdev committed:
Update on "Migrate reciprocal from the TH to Aten (CUDA)"


Benchmark (Debian buster, Release build, P400, turbo off):

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.reciprocal(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.reciprocal(a); torch.cuda.synchronize()', setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")', number=t))
```

Before:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.37665776500034553
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.3755079170005047
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45436807200167095
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7375995410002361
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.989992897000775
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7109333839998726
```

After:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.3880071029998362
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.38418854500014277
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45793338999919797
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7040607610006191
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.9805948499997612
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7015677830004279
```

[ghstack-poisoned]

xuhdev committed:
Update on "Migrate reciprocal from the TH to Aten (CUDA)"


Benchmark (Debian buster, Release build, P400, turbo off):

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.reciprocal(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.reciprocal(a); torch.cuda.synchronize()', setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")', number=t))
```

Before:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.37665776500034553
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.3755079170005047
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45436807200167095
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7375995410002361
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.989992897000775
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7109333839998726
```

After:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.3880071029998362
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.38418854500014277
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45793338999919797
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7040607610006191
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.9805948499997612
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7015677830004279
```

[ghstack-poisoned]

xuhdev committed:
Update on "Migrate reciprocal from the TH to Aten (CUDA)"


Benchmark (Debian buster, Release build, P400, turbo off):

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.reciprocal(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.reciprocal(a); torch.cuda.synchronize()', setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")', number=t))
```

Before:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.37665776500034553
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.3755079170005047
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45436807200167095
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7375995410002361
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.989992897000775
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7109333839998726
```

After:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.3880071029998362
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.38418854500014277
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45793338999919797
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7040607610006191
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.9805948499997612
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7015677830004279
```

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/cuda/UnaryOpsKernel.cu:
```diff
@@ -180,6 +180,15 @@ void trunc_kernel_cuda(TensorIterator& iter) {
 
+void reciprocal_kernel_cuda(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.dtype(), "reciprocal_cuda", [&]() {
+    using acc_t = std::conditional<std::is_same<scalar_t, at::Half>::value, float, scalar_t>::type;
```
You should use `using acc_t = acctype<scalar_t, /*is_cuda=*/true` to reuse our accumulate type logic: 

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/AccumulateType.h#L46

zou3519 reviewed (commented):
Code looks good, one suggestion

xuhdev committed:
Update on "Migrate reciprocal from the TH to Aten (CUDA)"


Benchmark (Debian buster, Release build, P400, turbo off):

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.reciprocal(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.reciprocal(a); torch.cuda.synchronize()', setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")', number=t))
```

Before:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.37665776500034553
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.3755079170005047
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45436807200167095
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7375995410002361
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.989992897000775
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7109333839998726
```

After:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.3880071029998362
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.38418854500014277
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45793338999919797
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7040607610006191
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.9805948499997612
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7015677830004279
```

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/cuda/UnaryOpsKernel.cu:
```diff
@@ -180,6 +181,15 @@ void trunc_kernel_cuda(TensorIterator& iter) {
 
+void reciprocal_kernel_cuda(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.dtype(), "reciprocal_cuda", [&]() {
+    using acc_t = acc_type<scalar_t, true>;
```
nit: I like providing a comment for boolean arguments (`acctype<scalar_t, /*is_cuda=*/true>`) to make it clearer

xuhdev committed:
Update on "Migrate reciprocal from the TH to Aten (CUDA)"


Benchmark (Debian buster, Release build, P400, turbo off):

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.reciprocal(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.reciprocal(a); torch.cuda.synchronize()', setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")', number=t))
```

Before:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.37665776500034553
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.3755079170005047
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45436807200167095
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7375995410002361
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.989992897000775
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7109333839998726
```

After:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.3880071029998362
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.38418854500014277
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45793338999919797
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7040607610006191
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.9805948499997612
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7015677830004279
```

[ghstack-poisoned]

xuhdev committed:
Update on "Migrate reciprocal from the TH to Aten (CUDA)"


Benchmark (Debian buster, Release build, P400, turbo off):

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.reciprocal(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.reciprocal(a); torch.cuda.synchronize()', setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")', number=t))
```

Before:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.37665776500034553
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.3755079170005047
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45436807200167095
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7375995410002361
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.989992897000775
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7109333839998726
```

After:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.3880071029998362
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.38418854500014277
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45793338999919797
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7040607610006191
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.9805948499997612
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7015677830004279
```

Differential Revision: [D18764186](https://our.internmc.facebook.com/intern/diff/D18764186)

[ghstack-poisoned]

xuhdev committed:
Update on "Migrate reciprocal from the TH to Aten (CUDA)"


Benchmark (Debian buster, Release build, P400, turbo off):

```python
import timeit

for n, t in [(10_000, 20000),
             (100_000, 20000)]:
    for dtype in ('torch.half', 'torch.float', 'torch.double'):
        print(f'torch.reciprocal(a) a.numel() == {n} for {t} times {dtype}')
        print(timeit.timeit(f'torch.reciprocal(a); torch.cuda.synchronize()', setup=f'import torch; a=torch.arange({n}, dtype={dtype}, device="cuda")', number=t))
```

Before:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.37665776500034553
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.3755079170005047
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45436807200167095
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7375995410002361
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.989992897000775
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7109333839998726
```

After:

```
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.half
0.3880071029998362
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.float
0.38418854500014277
torch.reciprocal(a) a.numel() == 10000 for 20000 times torch.double
0.45793338999919797
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.half
0.7040607610006191
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.float
0.9805948499997612
torch.reciprocal(a) a.numel() == 100000 for 20000 times torch.double
1.7015677830004279
```

Differential Revision: [D18764186](https://our.internmc.facebook.com/intern/diff/D18764186)

[ghstack-poisoned]

xuhdev commented:
This has actually been merged in 4ac614191ab35800799bff7fb6d5c6f992a16899 .
